kusto/kcsb.go

```diff
@@ -0,0 +1,485 @@
+	kcsb.ApplicationToken = appToken
+	return kcsb
+}
+
+// Creates a Kusto Connection string builder that will use existing authenticated az cli profile password.
+func (kcsb *ConnectionStringBuilder) WithAzCli() *ConnectionStringBuilder {
+	requireNonEmpty(dataSource, kcsb.DataSource)
+	kcsb.resetConnectionString()
+	kcsb.AzCli = true
+	return kcsb
+}
+
+/*
+Creates a Kusto Connection string builder that will authenticate with AAD application, using
+an application token obtained from a Microsoft Service Identity endpoint using user assigned id.
+*/
+func (kcsb *ConnectionStringBuilder) WithUserManagedIdentity(clientID string) *ConnectionStringBuilder {
+	requireNonEmpty(dataSource, kcsb.DataSource)
+	kcsb.resetConnectionString()
+	kcsb.MsiAuthentication = true
+	kcsb.ManagedServiceIdentity = clientID
+	return kcsb
+}
+
+/*
+Creates a Kusto Connection string builder that will authenticate with AAD application, using
+an application token obtained from a Microsoft Service Identity endpoint using system assigned id.
+*/
+func (kcsb *ConnectionStringBuilder) WithSystemManagedIdentity() *ConnectionStringBuilder {
+	requireNonEmpty(dataSource, kcsb.DataSource)
+	kcsb.resetConnectionString()
+	kcsb.MsiAuthentication = true
+	return kcsb
+}
+
+/*
+Creates a Kusto Connection string builder that will authenticate by launching the system default browser
+to interactively authenticate a user, and obtain an access token
+*/
+func (kcsb *ConnectionStringBuilder) WithInteractiveLogin(authorityID string) *ConnectionStringBuilder {
+	requireNonEmpty(dataSource, kcsb.DataSource)
+	kcsb.resetConnectionString()
+	if !isEmpty(authorityID) {
+		kcsb.AuthorityId = authorityID
+	}
+	kcsb.InteractiveLogin = true
+	return kcsb
+}
+
+/*
+Assigns ClientOptions to string builder that contains configuration settings like Logging and Retry configs for a client's pipeline.
+Read more at https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/azcore@v1.2.0/policy#ClientOptions
+*/
+func (kcsb *ConnectionStringBuilder) AttachPolicyClientOptions(options *azcore.ClientOptions) *ConnectionStringBuilder {
+	requireNonEmpty(dataSource, kcsb.DataSource)
+	if options == nil {
+		kcsb.ClientOptions = options
+	}
+	return kcsb
+}
+
+/*
+Create Kusto Conntection String that will be used for default auth mode. The order of auth will be via environment variables, managed identity and Azure CLI .
+Read more at https://learn.microsoft.com/azure/developer/go/azure-sdk-authentication?tabs=bash#2-authenticate-with-azure
+*/
+func (kcsb *ConnectionStringBuilder) WithDefaultAzureCredential() *ConnectionStringBuilder {
+	kcsb.resetConnectionString()
+	kcsb.DefaultAuth = true
+	return kcsb
+}
+
+// Method to be used for generating TokenCredential
+func (kcsb *ConnectionStringBuilder) newTokenProvider() (*TokenProvider, error) {
+	tkp := &TokenProvider{}
+	tkp.tokenScheme = BEARER_TYPE
+	// Fetches cloud meta data
+	fetchedCI, cierr := GetMetadata(context.Background(), kcsb.DataSource)
```

Comment:
Isn't this exactly what we wanted not to do?
This does the web request to cloud info when the client is created.
It should be done only when the token is requested.

Reply:
@AsafMah Only the first call for the new data source will be a web call. After that all will be cache access. 
the cloudSettingsInfo is required here to get tokenCredentials (defaulted with cloudSettinginfo)

Reply:
1. No, it's one call *per cluster*. Even if not, creating an object shouldn't do any long running actions.
Look what we do in all of the other SDKs.
2. Then lazily create them when needed, like in all other sdks.